src/sandstorm/grain.capnp

```diff
@@ -126,6 +126,77 @@ interface SandstormApi(AppObjectId) {
+  # in the future the event is scheduled, the more slack it has. You must explicitly set a non-zero
+  # value if you need better precision. Additionally, the value must be greater than
+  # `minimumSchedulingSlack`, or an exception will be thrown. If you need better precision than
+  # this, you should schedule a callback for slightly before the target time, then countdown the
+  # remaining time in-process.
+  #
+  # When the callback is called, Sandstorm will avoid killing the grain until the callback returns.
+  # If the callback throws a "disconnected" exception, or if Sandstorm is forced to kill the
+  # grain for some reason, the callback will be retried. After a limited number of failed retries,
+  # Sandstorm may alert the owner to a problem and stop retrying.
+
+  schedulePeriodic @12 (period :SchedulingPeriod, callback :Util.Runnable)
+                    -> (job :ScheduledJob);
+  # Schedules a callback to be called on a regular interval.
+  #
+  # Returns a persistable ScheduledJob handle that allows cancellation of the job.
+  #
+  # To activate the job, you must call confirm() on the returned `job` capability shortly after
+  # receiving it. Otherwise the job will remain inactive and may be automatically canceled.
+  #
+  # In order to improve resource utilization, Sandstorm prefers imprecise scheduling -- this allows
+  # it to choose a time of low activity to invoke the callback. Therefore, Sandstorm guarantees
+  # that the callback will be called once per scheduling period, but does not make any guarantees
+  # about exactly when in that period the call will occur. If you need more precise scheduling,
+  # you will need to use `scheduleAt()`.
+  #
+  # When the callback is called, Sandstorm will avoid killing the grain until the callback returns.
+  # If the callback throws a "disconnected" exception, or if Sandstorm is forced to kill the
+  # grain for some reason, the callback will be retried. After a limited number of failed retries,
+  # Sandstorm may alert the owner to a problem and stop retrying (but will call the callback again
+  # on the next scheduling period).
+}
+
+interface ScheduledJob @0xa34f0cfe24c69d74 {
+  # A handle to a job that has been scheduled by scheduleAt() or schedulePeriodic(). The
+  # job is not active until confirm() is called. This two-step activation handshake gives the
+  # grain a chance to save the ScheduledJob before activation, thus avoiding situations where
+  # a crash causes the job to be active but uncancellable.
+
+  confirm @0 ();
+  # Activates the job. This method only has an effect the first time it is called.
+
+  cancel @1 ();
+  # Cancels the job. This method only has an effect the first time it is called.
+}
+
+const minimumSchedulingSlack :Util.DurationInNs = 60000000000;
+# 60 seconds: The minimum value allowable for the `slack` parameter passed to `scheduleAt()`.
+
+enum SchedulingPeriod {
```

Comment:
This strikes me as not flexible enough for some use cases. Why not just provide a numeric duration?

Reply:
`If you need better precision than this, you should schedule a callback for slightly before the target time, then countdown the remaining time in-process.`

I presume this design is mostly large server/Oasis focused, in that you want to avoid the server being forced to run everyone's jobs precisely at midnight or the like. The initial boot of a grain presumably is higher load than it just running, so this staggers when grains start up?

Reply:
I'm talking about the restriction to hourly/daily/monthly/annually, not the `minimumSchedulingSlack` (which I think is fine). I don't think using a numeric duration for the period makes this any harder.

Reply:
Ah. Didn't see the lines below the comment in GitHub. Oops!

Reply:
It's been a while, but I think my goal here was to push apps to specify "rough" schedules rather than exact ones. It seems like most apps really only need rough schedules (like "invoke me once a day, but it doesn't really matter when"). My hope was to give the implementation some freedom to run these jobs at times when resource usage was low, or at least be able to spread out jobs so that it's not running tons of them at once.

If the developer can specify an exact period in seconds rather than use this enum, my worry is that developers will be more likely to assume that the scheduling will be precise. The enum at least hints that it's fuzzy.